snooty/types.py

```diff
@@ -66,10 +66,33 @@ def without_known_suffix(self) -> str:
         return fileid.as_posix()
 
 
-@dataclass
 class Diagnostic:
-    __slots__ = ("message", "severity", "start", "end")
+    # instance properties
```

Comment:
This is a bit of a refactor, but I actually think Diagnostic & all of its children should be moved to an `diagnostics.py` file, if you're up for that.